src/Discord.Net.Interactions/Info/ModalInfo.cs

```diff
@@ -0,0 +1,90 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace Discord.Interactions
+{
+    /// <summary>
+    ///     Represents a cached object initialization delegate.
+    /// </summary>
+    /// <param name="args">Property arguments array.</param>
+    /// <returns>
+    ///     Returns the constructed object.
+    /// </returns>
+    public delegate IModal ModalInitializer(object[] args);
+
+    /// <summary>
+    ///     Represents the info class of an <see cref="IModal"/> form.
+    /// </summary>
+    public class ModalInfo
+    {
+        internal readonly ModalInitializer _initializer;
+
+        /// <summary>
+        ///     Gets the title of this modal.
+        /// </summary>
+        public string Title { get; }
+
+        /// <summary>
+        ///     Gets the <see cref="IModal"/> implementation used to initialize this object.
+        /// </summary>
+        public Type Type { get; }
+
+        /// <summary>
+        ///     Gets a collection of the components of this modal.
+        /// </summary>
+        public IReadOnlyCollection<InputComponentInfo> Components { get; }
+
+        /// <summary>
+        ///     Gets a collection of the text components of this modal.
+        /// </summary>
+        public IReadOnlyCollection<TextInputComponentInfo> TextComponents { get; }
+
+        internal ModalInfo(Builders.ModalBuilder builder)
+        {
+            Title = builder.Title;
+            Type = builder.Type;
+            Components = builder.Components.Select(x => x switch
+            {
+                Builders.TextInputComponentBuilder textComponent => textComponent.Build(this),
+                _ => throw new InvalidOperationException($"{x.GetType().FullName} isn't a supported modal input component builder type.")
+            }).ToImmutableArray();
+
+            TextComponents = Components.Where(x => x is TextInputComponentInfo).Cast<TextInputComponentInfo>().ToImmutableArray();
```

Comment:
@CottageDwellingCat when you get the chance could you replace this mess with .OfType().ToImmutableArray() i forgot that metho even existed. sry im out of town and i dont have access to a pc